add hardened (Wolfi) images documentation for ECK (#6822)
elastic/docs-content: deploy-manage/deploy/cloud-on-k8s/hardened-images.md

---
navigation_title: Hardened (Wolfi) images
description: Learn how ECK uses Wolfi-based hardened images to reduce CVE exposure, and how to configure Wolfi images for Elastic Stack components.
applies_to:
  deployment:
    eck: all
products:
  - id: cloud-kubernetes
  - id: elasticsearch
  - id: kibana
  - id: elastic-agent
  - id: beats
  - id: apm
  - id: logstash
---

> # Hardened (Wolfi) images in {{eck}} [k8s-hardened-images]

Reviewer: we try to avoid brackets in titles. would either go with "Hardened Wolfi images" "Wolfi-hardened images" (I've seen both of these) or "Hardened images" 

to decide between these: do our users care more about the hardening or the technology used to do the hardening? if the former, skip "Wolfi" in the title

Reviewer 2: @jeanfabrice wdyt we say "Wolfi-based Hardened Docker images"

`Wolfi-based` appeared in [the blog - link](https://www.elastic.co/blog/reducing-cves-in-elastic-container-images#:~:text=The-,Wolfi%2Dbased,-images%20are%20not) we are referring to.

`Hardened Docker images` appeared in the [ES docs - link](https://www.elastic.co/guide/en/elasticsearch/reference/8.16/docker.html#docker-wolfi-hardened-image) too.

Also, IMHO,
* "Hardened docker image" is a thing, means more secure docker image
* "Wolfi-based" is a thing (more of a modifier), to explain the technology of how to achieve the "hardened docker image", which is wolfi-based.

> Elastic has partnered with [Chainguard](https://www.chainguard.dev/) to provide hardened container images based on [Wolfi](https://github.com/wolfi-dev/os), a minimal, security-focused Linux distribution designed for containerized environments. These images significantly reduce the CVE footprint of Elastic containers by including only the application and its necessary runtime dependencies. For background on this initiative, refer to the blog post [Reducing CVEs in Elastic container images](https://www.elastic.co/blog/reducing-cves-in-elastic-container-images).

Reviewer: our other wolfi docs use this link

```suggestion
Elastic has partnered with [Chainguard](https://www.chainguard.dev/) to provide hardened container images based on [Wolfi](https://wolfi.dev), a minimal, security-focused Linux distribution designed for containerized environments. These images significantly reduce the CVE footprint of Elastic containers by including only the application and its necessary runtime dependencies. For background on this initiative, refer to the blog post [Reducing CVEs in Elastic container images](https://www.elastic.co/blog/reducing-cves-in-elastic-container-images).
```

Reviewer 3: Is `Elastic has partnered with Chainguard` an official statement?

::::{note}

> Only images distributed via `docker.elastic.co` are officially supported by Elastic. Third-party hardened image sources, such as Docker Hardened Images (DHI) on Docker Hub, are not maintained by Elastic and fall outside the scope of Elastic support.

Reviewer: ```suggestion
Only images distributed through `docker.elastic.co` are officially supported by Elastic. Third-party hardened image sources, such as Docker Hardened Images (DHI) on Docker Hub, are not maintained by Elastic and fall outside the scope of Elastic support.
```

::::

> ## The ECK Operator image [k8s-hardened-images-operator]

Reviewer: ```suggestion
## The ECK operator image [k8s-hardened-images-operator]
```

> Since ECK **v2.15.0**, the ECK Operator image is built on Wolfi by default. No additional configuration is required — pulling the standard operator image from `docker.elastic.co` already provides a hardened, Wolfi-based container.

Reviewer: ```suggestion
Since ECK 2.15, the ECK operator image is built on Wolfi by default. No additional configuration is required —pulling the standard operator image from `docker.elastic.co` already provides a hardened, Wolfi-based container.
```

Reviewer 2: @jeanfabrice 

Do we want to also refer to our code base?

I can see in [the blog - link](https://www.elastic.co/blog/reducing-cves-in-elastic-container-images#:~:text=Looking%20at%20the%20ECK%20repository%20codebase%20and%20especially%20the%20Dockerfile%2C%20it%20illustrates%20the%20best%20practices%20mentioned%20above), we refer to the code base.

I guess it might be friendly to user if we can show them the codebase - https://github.com/elastic/cloud-on-k8s/blob/v2.15.0/build/Dockerfile#L30, like we do in blog.

But I don't have strong opinion, and will leave the decision to you. :)

Reviewer 2: @jeanfabrice @shainaraskas one more thing, do we want to say ECK operator ubi image also uses wolfi, per https://github.com/elastic/cloud-on-k8s/blob/v2.15.0/build/Dockerfile-ubi?

(Because @jeanfabrice we have "This flag cannot be combined with the `--ubi-only` flag." for stack images, my personal feeling is, logically thinking, users may want to know how ECK is handling ubi images, with wolfi or not).

Reviewer: the ubi thing makes sense to me 

referring to the code base might not be necessary because the updates under the hood are more of an implementation detail. users should be able to trust that we are keeping our dependencies up to date without us calling it out specifically.

Reviewer 4: Chiming in here

> @jeanfabrice @shainaraskas one more thing, do we want to say ECK operator ubi image also uses wolfi, per https://github.com/elastic/cloud-on-k8s/blob/v2.15.0/build/Dockerfile-ubi?

the above statement doesn't feel right. Looking [here](https://github.com/elastic/cloud-on-k8s/blob/main/build/Dockerfile#L97). Sure there is an intermediate builder image step that just builds the operator binary. Then for the final ubi we copy just the binary and uses the ubi base image. The end result is an ubi-based image that has nothing to do with wolfi.

Reviewer 2: @pkoutsovasilis thanks. Is it a bug in ECK that we say it defaults to wolfi, but actually for ubi flavor, it is not?

Reviewer 4: I am not sure I follow 😆 The default is the static target in the Dockerfile, linked above, which is indeed Wolfi-based, so what's the bug?! Additionally, in the same Dockerfile we define the target for ubi. To this end, replying to your previous point this doesn't make the ubi image use wolfi, these are separate images with different base

> ## {{stack}} images managed by {{eck}} [k8s-hardened-images-stack]

Reviewer: how does autoops fit in? looks like that's wolfi by default
https://github.com/elastic/cloud-on-k8s/blob/main/pkg/controller/common/container/container.go#L69

> ## {{stack}} images managed by {{eck}} [k8s-hardened-images-stack]

Reviewer: ```suggestion
## {{stack}} images managed by ECK [k8s-hardened-images-stack]
```

> Wolfi-based variants of the {{stack}} images ({{es}}, {{kib}}, {{agent}}, {{beats}}) are available from v8.16.0 onwards (v8.15.0 for {{apm-server}} and {{ls}}). However, ECK does not pull Wolfi variants by default for {{stack}} components — the standard images are used unless explicitly overridden.

Reviewer: ```suggestion
Wolfi-based variants of the {{stack}} images ({{es}}, {{kib}}, {{agent}}, {{beats}}) are available from version 8.16 onwards (8.15 for {{apm-server}} and {{ls}}). However, ECK does not pull Wolfi variants by default for {{stack}} components — the standard images are used unless explicitly overridden.
```

Reviewer 2: @shainaraskas should we do sth like a more human readable list?
It seems like we have many images.
https://github.com/elastic/cloud-on-k8s/blob/main/pkg/controller/common/container/container.go#L55-L70

Also per our (public) KB:
https://support.elastic.co/knowledge/662e34c1

My personal vote would be something like we have in KB:

- APM Server & Logstash: 8.15.0+
- Elasticsearch, Kibana, Elastic Agent, Beats (e.g. Filebeat, Metricbeat, Auditbeat, Packetbeat, Heartbeat, etc.), & Enterprise Search: 8.16.0+

---

But again, this is just a cosmetic thing and I don't have strong opinion :)

Reviewer: that would be fine, but I'd skip the beats list

```suggestion
Wolfi-based variants of the following {{stack}} images are available:

* {{stack}} version 8.16+: {{es}}, {{kib}}, {{agent}}, {{beats}}
* {{stack}} version 8.15+: {{apm-server}} and {{ls}})

ECK does not pull Wolfi variants by default for {{stack}} components — the standard images are used unless explicitly overridden.
```

To have the operator pull Wolfi-based images natively for the {{stack}} components it manages, set the `container-suffix` configuration flag to `-wolfi`. The operator then appends this suffix to the container images it resolves. This flag cannot be combined with the `--ubi-only` flag. For more details, refer to [ECK configuration flags](cloud-on-k8s://reference/eck-configuration-flags.md).

> For example, set the suffix in the `elastic-operator` ConfigMap:

Reviewer: ```suggestion
For example, set the suffix in the `elastic-operator` ConfigMap.

::::{warning}
The `container-suffix` flag applies to **every** resource the operator manages, including [{{package-registry}}](package-registry.md) and [{{hosted-ems}}](elastic-maps-server.md). Both are already built on Wolfi and don't publish `-wolfi` variants, so setting `container-suffix: -wolfi` without also overriding `.spec.image` for these resources will cause their image pulls to fail.

If your deployment includes {{package-registry}} or {{hosted-ems}}, apply the overrides in [Override images for {{package-registry}} and {{hosted-ems}}](#k8s-hardened-images-prebuilt) at the same time as the `container-suffix` change.
::::

```

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: elastic-operator
  namespace: elastic-system
data:
  eck.yaml: |-

>     container-suffix: -wolfi

Reviewer: how is this done with helm?

```

> ::::{warning}

Reviewer: this warning is very long and should be broken up. putting a couple other comments on to show the strategy I recommend

Reviewer 2: I get your point @shainaraskas, but my vote is to leave it there as a whole piece, since this is really an important one. Not following this correctly will cause image not being pulled correctly.

This is a product level gap, and if we can have 
* ECK operator flag to enable wolfi stack images, 
   * and inside the operator logic, 
      * it automatically and smartly understand what images are default to wolfi, 
      * what are required to add `container-suffix: -wolfi` to apply wolfi, 

Then we can significantly simplify this block as a simple sentence - use that flag.
Before we get there, I would prefer we have this warning significant enough to raise good awareness.

Reviewer: there should be a warning still, but the workaround should not be inside of the warning. my other suggestions cpature this.

> ::::{warning}

Reviewer: do EPR and maps-server install pages need a warning about this?

Reviewer 2: ++ on having a section in those two pages about wolfi, mention "wolfi image enabled by default for the two products", and when use in ECK, if user specify `container-suffix: -wolfi` to run wolfi on other stack images, they need to pay addition attention to follow what @jeanfabrice wrote in the warning block.

<img width="763" height="906" alt="Image" src="https://github.com/user-attachments/assets/13c55fb8-acbb-4717-b735-d9d0363b6a6c" />

> ::::{warning}
> The `container-suffix` flag applies to **all** resources managed by the operator, including [{{package-registry}}](package-registry.md) and [{{hosted-ems}}](elastic-maps-server.md). The images for these two components are **already natively based on Wolfi** and do not have a `-wolfi` suffix. With `container-suffix: -wolfi` set, the operator would try to pull non-existent suffixed images, causing the image pull to fail.
> 
> For these resources, explicitly set the container image in the manifest using `.spec.image` (without the suffix).
> 
> For [{{package-registry}}](package-registry.md), use:
> 
> ```yaml subs=true
> apiVersion: packageregistry.k8s.elastic.co/v1alpha1
> kind: PackageRegistry
> metadata:
>   name: package-registry-sample
>   namespace: default
> spec:
>   count: 1
>   version: {{version.stack}}
>   image: docker.elastic.co/package-registry/distribution:{{version.stack}}
> ```
> 
> Other valid tags include `lite-<version>`, `production` and `lite`. For more details, refer to [Air-gapped environments](/reference/fleet/air-gapped.md#air-gapped-diy-epr).
> 
> For [{{hosted-ems}}](elastic-maps-server.md), use:
> 
> ```yaml subs=true
> apiVersion: maps.k8s.elastic.co/v1alpha1
> kind: ElasticMapsServer
> metadata:
>   name: quickstart
> spec:
>   count: 1
>   version: {{version.stack}}
>   image: docker.elastic.co/elastic-maps-service/elastic-maps-server:{{version.stack}}
> ```
> 
> ::::

Reviewer: ````suggestion
## Override images for {{package-registry}} and {{hosted-ems}} [k8s-hardened-images-prebuilt]

{{package-registry}} and {{hosted-ems}} don't publish `-wolfi` image variants because they're already built on Wolfi. When `container-suffix: -wolfi` is set, override `.spec.image` on each of these resources to use the unsuffixed image instead.

### {{package-registry}} [k8s-hardened-images-epr]

Set `.spec.image` in your `PackageRegistry` manifest without the `-wolfi` suffix:

```yaml subs=true
apiVersion: packageregistry.k8s.elastic.co/v1alpha1
kind: PackageRegistry
metadata:
  name: package-registry-sample
  namespace: default
spec:
  count: 1
  version: {{version.stack}}
  image: docker.elastic.co/package-registry/distribution:{{version.stack}}
```

Other valid tags include `lite-<version>`, `production`, and `lite`. For more details, refer to [Air-gapped environments](/reference/fleet/air-gapped.md#air-gapped-diy-epr).

### {{hosted-ems}} [k8s-hardened-images-ems]

Set `.spec.image` in your `ElasticMapsServer` manifest without the `-wolfi` suffix:

```yaml subs=true
apiVersion: maps.k8s.elastic.co/v1alpha1
kind: ElasticMapsServer
metadata:
  name: quickstart
spec:
  count: 1
  version: {{version.stack}}
  image: docker.elastic.co/elastic-maps-service/elastic-maps-server:{{version.stack}}
```
````

> ## Further reading [k8s-hardened-images-further-reading]
> 
> - [Reducing CVEs in Elastic container images](https://www.elastic.co/blog/reducing-cves-in-elastic-container-images)

Reviewer: we link this at the very top so we can probably kill it 
```suggestion
```
